ska3-core to use pytest < 8.0.0

During the release of 2024.1, there was a problem with pytest version `8.0.0` and later. It was decided to downgrade the pytest version for now.

The problem stemmed from passing the `-c` option to pytest.

The following succeeds:
```
cd /export/jgonzale/github-workflows/miniconda3/envs/ska3-matlab-2024.2rc3/lib/python3.11/site-packages
pytest Quaternion -c /export/jgonzale/ska_testr/pytest.ini
```
whereas this fails
```
cd /export/jgonzale/github-workflows/miniconda3/envs/ska3-matlab-2024.2rc3/lib/python3.11/site-packages
pytest Quaternion -c /proj/sot/ska/jgonzalez/git/ska_testr/pytest.ini
```

We believe that the behavior has to do with pytest setting the root to the location of `pytest.ini`. We have not check further why it fails in one case and not the other.

The original error is pasted below:
```
(ska3-masters) jeanconn-fido> run_testr --include Quaternion
****************************************
*** package Quaternion               ***
****************************************

Copying input tests /proj/sot/ska/jeanproj/git/ska_testr/packages/Quaternion to output dir /proj/sot/ska/jeanproj/git/ska_testr/outputs/logs/Linux_2024-03-01T16-46-36_40e13bc_fido.cfa.harvard.edu/Quaternion
Running python test_unit.py script
============================= test session starts ==============================
platform linux -- Python 3.11.8, pytest-8.0.2, pluggy-1.4.0 -- /fido.real/conda/envs/ska3-masters/bin/python
rootdir: /proj/sot/ska/jeanproj/git/ska_testr
configfile: pytest.ini
plugins: timeout-2.2.0, anyio-4.3.0
collecting ... ERROR: found no collectors for /fido.real/conda/envs/ska3-masters/lib/python3.11/site-packages/Quaternion

collected 0 items / 1 error

==================================== ERRORS ====================================
________________________ ERROR collecting test session _________________________
pluggy/_hooks.py:501: in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
pluggy/_manager.py:119: in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
_pytest/python.py:211: in pytest_collect_directory
    if pkginit.is_file():
../pathlib.py:1267: in is_file
    return S_ISREG(self.stat().st_mode)
../pathlib.py:1013: in stat
    return os.stat(self, follow_symlinks=follow_symlinks)
E   PermissionError: [Errno 13] Permission denied: '/fido.real/lost+found/__init__.py'
- generated xml file: /proj/sot/ska/jeanproj/git/ska_testr/outputs/logs/Linux_2024-03-01T16-46-36_40e13bc_fido.cfa.harvard.edu/Quaternion/test_unit.py.xml -
=========================== short test summary info ============================
ERROR ../../../../../../../proj/sot/ska/jeanproj/git/ska_testr/::fido.real - ...
=============================== 1 error in 0.26s ===============================
Traceback (most recent call last):
  File "/proj/sot/ska/jeanproj/git/ska_testr/outputs/logs/Linux_2024-03-01T16-46-36_40e13bc_fido.cfa.harvard.edu/Quaternion/test_unit.py", line 2, in <module>
    testr.testr()
  File "/fido.real/conda/envs/ska3-masters/lib/python3.11/site-packages/testr/runner.py", line 51, in testr
    return test(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/fido.real/conda/envs/ska3-masters/lib/python3.11/site-packages/testr/runner.py", line 208, in test
    raise TestError('Failed')
testr.runner.TestError: Failed
Running python post_check_logs.py script
Traceback (most recent call last):
  File "/proj/sot/ska/jeanproj/git/ska_testr/outputs/logs/Linux_2024-03-01T16-46-36_40e13bc_fido.cfa.harvard.edu/Quaternion/post_check_logs.py", line 3, in <module>
    check_files('test_*.log', ['warning', 'error'])
  File "/fido.real/conda/envs/ska3-masters/lib/python3.11/site-packages/testr/packages.py", line 691, in check_files
    raise ValueError('Found matches in check_files:\n{}'.format('\n'.join(matches)))
ValueError: Found matches in check_files:
'error' matched at test_unit.log:5 :: collecting ... ERROR: found no collectors for /fido.real/conda/envs/ska3-masters/lib/python3.11/site-packages/Quaternion
'error' matched at test_unit.log:7 :: collected 0 items / 1 error
'error' matched at test_unit.log:9 :: ==================================== ERRORS ====================================
'error' matched at test_unit.log:10 :: ________________________ ERROR collecting test session _________________________
'error' matched at test_unit.log:21 :: E   PermissionError: [Errno 13] Permission denied: '/fido.real/lost+found/__init__.py'
'error' matched at test_unit.log:24 :: ERROR ../../../../../../../proj/sot/ska/jeanproj/git/ska_testr/::fido.real - ...
'error' matched at test_unit.log:25 :: =============================== 1 error in 0.26s ===============================
'error' matched at test_unit.log:33 :: raise TestError('Failed')
'error' matched at test_unit.log:34 :: testr.runner.TestError: Failed
```

--- a/testr/runner.py
+++ b/testr/runner.py
@@ -185,7 +185,17 @@ def test(*args, **kwargs):
     pkg_paths = [os.path.dirname(calling_func_file)] + ['..'] * len(pkg_names)
     pkg_dir = os.path.join(*pkg_names)
 
-    with chdir(os.path.join(*pkg_paths)):
+    # Set the rootdir to the top of the package directory. This gets used in two ways:
+    # 1. To set the working directory for the test run.
+    # 2. To set the default for the --rootdir option in the pytest call.
+    #
+    # For part 2, see https://github.com/sot/skare3/issues/1294 for context, but in
+    # summary pytest > 8.0.0 uses the location of a pytest.ini file to set the rootdir
+    # unless it is explicitly specified. This is generally NOT what we want.
+    rootdir = os.path.abspath(os.path.join(*pkg_paths))
+    kwargs.setdefault('rootdir', rootdir)
+
+    with chdir(rootdir):
         if with_coverage:
             coverage_file = os.path.join(
                 os.environ['TESTR_OUT_DIR'],

--- a/testr/runner.py
+++ b/testr/runner.py
@@ -68,7 +68,9 @@ def test(*args, **kwargs):
     :param package_from_dir: set package name from parent directory name (default=False)
     :param verbose: run pytest in verbose (-v) mode (default=False)
     :param show_output: run pytest in show output (-s) mode (default=False)
-    :param \*\*kwargs: additional keyword args to pass to pytest
+    :param \*\*kwargs: additional command line options to pass to pytest, where the
+        key is the option name (e.g. "--log-level") and the value is the option value.
+        This is passed to the pytest CLI as the string ``{key}={value}``.
 
     :returns: number of test failures
     """
@@ -193,7 +195,8 @@ def test(*args, **kwargs):
     # summary pytest > 8.0.0 uses the location of a pytest.ini file to set the rootdir
     # unless it is explicitly specified. This is generally NOT what we want.
     rootdir = os.path.abspath(os.path.join(*pkg_paths))
-    kwargs.setdefault('rootdir', rootdir)
+    kwargs.setdefault('--rootdir', rootdir)
+    args += tuple([f'{k}={v}' for k, v in kwargs.items()])
 
     with chdir(rootdir):
         if with_coverage:
@@ -206,13 +209,13 @@ def test(*args, **kwargs):
                 f'--rcfile={coverage_config}',
                 f'--data-file={coverage_file}',
                 '-m', 'pytest', pkg_dir
-            ] + list(args) + [f'{k}={v}' for k, v in kwargs]
+            ] + list(args)
             with stdout_context():
                 process = subprocess.run(cmd, stdout=sys.stdout, stderr=subprocess.STDOUT)
             rc = process.returncode
         else:
             with stdout_context():
-                rc = pytest.main([pkg_dir] + list(args), **kwargs)
+                rc = pytest.main([pkg_dir] + list(args))
 
     if rc and raise_exception:
         raise TestError('Failed')

